ex-3-recursion/look-for-key.js

```diff
@@ -1,21 +1,22 @@
 function lookForKey( arr ){
+    let result = "not found";
 
-    for (let i = 0; i < arr.length; i++) {
+    for ( let i = 0; i < arr.length; i++ ) {
         if( arr[i] == "key" ){
             console.log( "Key found.");
-            //return "Key found.";   - how to return when found?
+            result = "Key found";
         } else if( Array.isArray( arr[i] ) ){
-            lookForKey( arr[i] );
+            result = lookForKey( arr[i] );
         }
     }
+
+    return result;
 }
 
 let arr = [
     ["stuff", "things"],
     ["stuff", "things"],
-    ["stuff", "things", "key"],
-    ["stuff", "things"],
+    ["stuff", "things", "key"]
```

Comment:
Looks great. Alternatively you could simply return at each of the places where you assigned to result and then returned "not found" at the end 👍🏼 

```js

function lookForKey( arr ){
    for ( let i = 0; i < arr.length; i++ ) {
        if( arr[i] == "key" ){
            console.log( "Key found.");
            return "Key found";
        } else if( Array.isArray( arr[i] ) ){
            return lookForKey( arr[i] );
        }
    }

    return "not found";
}
```

Reply:
Thanks for the tips!  I meant to leave a comment last night on this, as my update didn't really fix much.  It works for the examples:

```
["key"]

[["key"]]
```

I'm still hitting walls on this one.  My latest update works until there's another array following the array with the key.  I've been working with your suggestion above instead.  But the following is failing:

```
let arr = [
    ["stuff", "things"],
    ["stuff", "things"],
    ["stuff", "things", "key"]
];
```

It seems to only check the first array in this case, then returns not found.  I **know** I'm missing something simple here.

Reply:
Oh! It's tricky but I think I know what the issue is. If we always use the value from the inner `lookForKey` we lose something in certain cases.

The simplest case I can think of to cause the problem is this:

```
[
  ["foo"],
  ["key"],
]
```

What happens if we call `lookForKey` and it returns `not found`? Do we need to always use that value, or should we make sure that we're checking the array exhaustively before returning not found?

This is kind of complicated to explain and I'm trying not to give away too much details of how to fix this. 

One tool that might be helpful is to `console.log` the "interesting" variable for each iteration of the loop. I would start by logging `arr`, `arr[i]` and `lookForKey(arr[i])` on each iteration and seeing if you can find a pattern for where the breaking happens.

Reply:
Thanks!  On it.

Reply:
I spent an embarrassing amount of time on this, but I had to find the solution.  Finally working for any multi-dimensional array: https://github.com/mdmoore/grokking-algorithms/pull/5/commits/18be44cf2ec065f1a25143b730bdda909b51f85b 🎉